src/mito2/src/schedule/scheduler.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU8, Ordering};
+use std::sync::Arc;
+
+use common_telemetry::info;
+use snafu::{ensure, ResultExt};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use tokio_util::sync::CancellationToken;
+
+use crate::error::{
+    InvalidFlumeSenderSnafu, InvalidSchedulerStateSnafu, Result, StopSchedulerSnafu,
+};
+
+pub type Job = Pin<Box<dyn Future<Output = ()> + Send>>;
+
+///The state of scheduler
+const STATE_RUNNING: u8 = 0;
+const STATE_STOP: u8 = 1;
+const STATE_AWAIT_TERMINATION: u8 = 2;
+
+/// The consumer count
+const CONSUMER_NUM: u8 = 2;
+
+/// [Scheduler] defines a set of API to schedule Jobs
+#[async_trait::async_trait]
+pub trait Scheduler {
+    /// Schedules a Job
+    async fn schedule(&self, req: Job) -> Result<()>;
+
+    /// Stops scheduler. If `await_termination` is set to true, the scheduler will wait until all tasks are processed.
+    async fn stop(&self, await_termination: bool) -> Result<()>;
+}
+
+/// Request scheduler based on local state.
+pub struct LocalScheduler {
+    /// Sends jobs to flume bounded channel
+    sender: flume::Sender<Job>,
+    /// Task handles
+    handles: tokio::sync::Mutex<Vec<JoinHandle<()>>>,
+    /// Token used to halt the scheduler
+    cancel_token: CancellationToken,
+    /// State of scheduler
+    state: Arc<AtomicU8>,
+}
+
+impl LocalScheduler {
+    /// cap: flume bounded cap
+    /// rev_num: the number of bounded receiver
+    pub fn new(cap: usize, rev_num: usize) -> Self {
+        let (tx, rx) = flume::bounded(cap);
+        let token = CancellationToken::new();
+        let state = Arc::new(AtomicU8::new(STATE_RUNNING));
+
+        let mut handles = Vec::with_capacity(rev_num);
+
+        for _ in 0..rev_num {
+            let child = token.child_token().clone();
```

Comment:
```suggestion
            let child = token.child_token();
```

There's a slight difference between [cloning a token](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#method.clone) and [creating a child token](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#method.child_token).